medusajs/medusa
# feat: Add DiscountConditions

**What**
Introduces DiscountConditions and removes`valid_for`

**Why**
To enable more advanced promotions such as creating a discount for a specific set of products, collections, types, tags, and customer groups.

**How**
Introduce new entity DiscountCondition that holds a type ("products"), operator ("in" meaning resources must be in the condition for the discount to be valid), and a reference to a specific type of DiscountCondition e.g. DiscountConditionProduct. Read more about the entity implementation [here](https://github.com/medusajs/medusa/discussions/1148). #1146 

Convert DiscountService to TypeScript. #1149 

Include new conditions in payload to update and create discount. #1170 

Add DiscountCondition repository that allows us to add, update and remove conditions in bulk, and check the applicableness of a given line item or customer. #1156, #1172, #1212 

Refactor the discount calculation for a LineItem from TotalsService to DiscountService. #1224

Add discount condition factory for integration tests. 

**Testing**
Unit tests
- Invalidate discount when no customer id is provided
- Invalidate discount when no customer has no groups
- Return true if groups exist on customer (assuming the group passes the check in repository layer)
- Calculate fixed discount applied on total amount
- Calculate fixed discount applied on an item basis
- Calculate percentage discount
- Return full line item price in case discount exceeds this
- Return early if item is not discountable

_Integration tests_ (added to existing test suite)
- Create discount with conditions using factory
- Create discount with conditions
- Create discount with conditions and update same conditions in a subsequent call
- Fails to create discount when multiple condition types are provided in payload
- Fails to update discount when multiple condition types are provided in payload
- Fails to add condition on rule with existing comb. of type and operator
